crates/services/producer/src/block_producer.rs

```diff
@@ -404,7 +409,17 @@ where
             if total_cost > gas_limit {
                 break;
             } else {
-                new_best = DaBlockHeight(height);
+                let transactions_number = self
```

Comment:
`get_cost_for_block` and `get_transactions_number_for_block` do the same lookup into the database. I think it will be better to have one function that returns both values for performance reasons.